File: python/백준/부르트 포스/30804 과일 탕후루.py
def max_fruit_length(N, fruits):
    from collections import defaultdict
    
    left = 0
    fruit_count = defaultdict(int)  # 과일의 개수를 저장하는 딕셔너리
    max_length = 0
    
    for right in range(N):
        fruit_count[fruits[right]] += 1  # 오른쪽 과일 추가
        
        # 과일 종류가 3개 이상이면 왼쪽 포인터 이동
        while len(fruit_count) > 2:
            fruit_count[fruits[left]] -= 1
            if fruit_count[fruits[left]] == 0:
                del fruit_count[fruits[left]]  # 과일 개수가 0이면 제거
            left += 1  # 왼쪽 포인터 이동
        
        # 현재 윈도우 크기 갱신
        max_length = max(max_length, right - left + 1)
    
    return max_length

# 입력 받기
N = int(input())
fruits = list(map(int, input().split()))

# 결과 출력
print(max_fruit_length(N, fruits))

File: cpp/백준/투포인터/30804 과일 탕후루.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int N;
    cin >> N;
    vector<int> fruits(N);
    for (int i = 0; i < N; ++i) cin >> fruits[i];

    int left = 0;
    unordered_map<int, int> fruit_count;
    int max_length = 0;

    for (int right = 0; right < N; ++right) {
        fruit_count[fruits[right]] += 1;
        while ((int)fruit_count.size() > 2) {
            fruit_count[fruits[left]] -= 1;
            if (fruit_count[fruits[left]] == 0) {
                fruit_count.erase(fruits[left]);
            }
            left += 1;
        }
        max_length = max(max_length, right - left + 1);
    }

    cout << max_length << "\n";
    return 0;
}
